FlowCrypt/src/main/java/com/flowcrypt/email/jetpack/viewmodel/RefreshPrivateKeysFromEkmViewModel.kt

```diff
@@ -148,54 +155,56 @@ class RefreshPrivateKeysFromEkmViewModel(application: Application) : AccountView
     val keysStorage = KeysStorageImpl.getInstance(context)
     val existingKeyEntities = keysStorage.getRawKeys()
     val existingPgpKeyDetailsList = keysStorage.getPgpKeyDetailsList()
-    var passphrase: Passphrase? = null
+    val keysToUpdate = mutableListOf<PgpKeyDetails>()
+    val keysToAdd = mutableListOf<PgpKeyDetails>()
 
     for (fetchedPgpKeyDetails in fetchedPgpKeyDetailsList) {
-      if (!fetchedPgpKeyDetails.usableForEncryption) {
-        //ask Tom
-        //we skip a key that is not usable for encryption
-        continue
+      val existingPgpKeyDetails = existingPgpKeyDetailsList.firstOrNull {
+        it.fingerprint == fetchedPgpKeyDetails.fingerprint
       }
 
-      val existingPgpKeyDetails =
-        existingPgpKeyDetailsList.firstOrNull { it.fingerprint == fetchedPgpKeyDetails.fingerprint }
-
       if (existingPgpKeyDetails != null) {
         if (fetchedPgpKeyDetails.isNewerThan(existingPgpKeyDetails)) {
-          if (passphrase == null) {
-            passphrase = getUsablePassphraseFromCache()
-          }
-
-          val existingKeyEntity = existingKeyEntities.first {
-            it.fingerprint == existingPgpKeyDetails.fingerprint
-          }
-
-          val safeVersionOfPrvKey = protectAndEncryptInternally(passphrase, fetchedPgpKeyDetails)
-          roomDatabase.keysDao().updateSuspend(
-            existingKeyEntity.copy(
-              privateKey = safeVersionOfPrvKey,
-              publicKey = fetchedPgpKeyDetails.publicKey.toByteArray()
-            )
-          )
+          keysToUpdate.add(fetchedPgpKeyDetails)
         }
       } else {
-        if (passphrase == null) {
-          passphrase = getUsablePassphraseFromCache()
-        }
+        keysToAdd.add(fetchedPgpKeyDetails)
+      }
+    }
+
+    if (keysToUpdate.isEmpty() && keysToAdd.isEmpty()) {
+      return@withContext
+    }
+
+    val passphrase: Passphrase = getUsablePassphraseFromCache()
 
-        val safeVersionOfPrvKey = protectAndEncryptInternally(passphrase, fetchedPgpKeyDetails)
-        val keyEntity = fetchedPgpKeyDetails.toKeyEntity(activeAccount).copy(
+    for (pgpKeyDetails in keysToUpdate) {
+      val existingKeyEntity = existingKeyEntities.first {
+        it.fingerprint == pgpKeyDetails.fingerprint
+      }
+
+      val safeVersionOfPrvKey = protectAndEncryptInternally(passphrase, pgpKeyDetails)
+      roomDatabase.keysDao().updateSuspend(
+        existingKeyEntity.copy(
           privateKey = safeVersionOfPrvKey,
-          storedPassphrase = null
+          publicKey = pgpKeyDetails.publicKey.toByteArray()
         )
-        roomDatabase.keysDao().insertSuspend(keyEntity)
-      }
+      )
```

Comment:
https://github.com/FlowCrypt/flowcrypt-android/blob/e336f06c1872ee644a49cfe695b8b3d1ae1d59e5/FlowCrypt/src/main/java/com/flowcrypt/email/database/entity/KeyEntity.kt#L43

We don't have such fields. Actually, I'm thinking to delete @ColumnInfo(name = "public_key") val publicKey: ByteArray to prevent errors. For now, we parse private keys with PGPainless and use this info to check the key details(including a public key from the private key)

Reply:
I wonder if there is performance penalty to recreate public key from private key, instead of reading public key directly (in case you only need the public key that moment)

Reply:
ok. In that case, I suppose you could remove the public part from storage.